[Bundle] Add mechanism to archive additional objects to bundle

mciprian13 opened the pull request:
**Summary**
Added mechanism to archive additional objects to bundle:
- Added a CMake utility function to serialize files and add to Glow binary files
- Enabled this mechanism for the CPUBackend as an example for other backends to use this mechanism (either public or private backends).
- All the `.o` objects located in `glow/lib/Backends/CPU/libobj_cpu` will be automatically serialized and added in a variable called `cpuObjectRegistry` with memory buffers as `llvm::MemoryBufferRef` at glow build-time
- The command line option `-bundle-objects` controls which of the serialized objects will be archived together with the bundle
- The objects can also be chosen dynamically at runtime by using the function: `void LLVMIRGen::addBundleObject()`

**Documentation**
Short doc in the folder where objects should be located.

**Test Plan**
Added unit test in RELEASE_WITH_EXPENSIVE_TESTS.

opti-mix commented on lib/LLVMIRCodeGen/libobj/Readme.md:
```diff
@@ -0,0 +1,19 @@
+# Archiving additional objects within the bundle object
+
+All the object files (.o) stored in this directory will be automatically
```
LLVMIRCodeGen is a reusable library component used by different LLVMBackend-derived backends. Those backends may generate code for different targets and architectures. Thus we putting target-specific objects files into such a shared directory does not make a lot of sense, IMHO. If at all, they should be placed in the backend specific directories.

mciprian13 replied:
Yeah maybe we could change the location of this directory. Since there is a strict convention about the location of the backends, we could choose to select all the files located in the `libobj` sub-directory from each backend directory (if exists), for example `glow/lib/Backends/CPU/libobj`. Although I would also keep this directory for resources which could be common to all backends (e.g. object files which contain data only but no code).

opti-mix commented on lib/LLVMIRCodeGen/libobj/Readme.md:
```diff
@@ -0,0 +1,19 @@
+All the object files (.o) stored in this directory will be automatically
+serialized as C buffer arrays and stored within Glow library at build time.
+When building bundles with Glow tools the user can opt to archive additional
+objects within the bundle object file by specifying the following command line
```
Why do you consider only object files and not e.g. some other resource files that maybe needed by the bundle?

mciprian13 replied:
The files must be object files because the intention is to archive them together with the original bundle created by Glow into a static library (object archive). Are there any file types I should consider?

opti-mix commented on lib/LLVMIRCodeGen/CommandLine.cpp:
```diff
@@ -108,3 +108,8 @@ llvm::cl::opt<bool> bundleAPIVerbose(
+
+llvm::cl::list<std::string> bundleObjectsOpt(
+    "bundle-objects",
+    llvm::cl::desc("Option to add other external object files to the bundle"),
```
Do you expect list of file names here? If so, maybe state it clearly.

mciprian13 replied:
I changed the description to state that we expect a list of names.

opti-mix commented on lib/LLVMIRCodeGen/BundleSaver.cpp:
```diff
@@ -450,6 +457,66 @@ void BundleSaver::emitSymbolTable() {
+    llvm::StringRef bundlePath, const std::vector<std::string> &bundleObjects) {
+
+  // If we do not have extra object files then return early.
+  if (!bundleObjects.size()) {
```
Use `if(bundleObjects.empty())`?

mciprian13 replied:
Done.

opti-mix commented on lib/LLVMIRCodeGen/BundleSaver.cpp:
```diff
@@ -450,6 +457,66 @@ void BundleSaver::emitSymbolTable() {
                            irgen_->getBundleName() + "SymbolTable");
 }
 
+void BundleSaver::addBundleObjects(
```
So, you try to produce a static library archive from the original object file generated by Glow and some additional object files, right?

mciprian13 replied:
That is correct.

opti-mix commented on lib/LLVMIRCodeGen/BundleSaver.cpp:
```diff
@@ -450,6 +457,66 @@ void BundleSaver::emitSymbolTable() {
+    }
+    // Find current object and add it as archive member.
+    bool objectFound = false;
+    for (const auto &memBuffRef : bundleObjectRegister) {
```
Why does bundleObjectRegister need to be produced before Glow itself is built (if I understand your PR correctly and how it produces BundleObjectRegister.h). Why do you need all this trouble with CMake? Why not read objects file at runtime, when BundleSaver is doing its work? In general, I'm not a big fan of this CMake magic. If there less build-system based solutions, it would be good to hear about them. Ideally, no build system tricks should be needed at all and everything could be done when BundleSaver runs. This way it will be also much easier to customize the behavior through command-line arguments and parameters passed by backends.

mciprian13 replied:
The intention is to embed in Glow binary some object files which are required for some custom LLVM backends (you can look into the usecase described below) in order to generate a self contained bundle with some extra dependencies which could be hidden to the end-user.
In my opinion there is no much value in having a CLI arg which describes some paths to some **object files external to the Glow tools** because those can be linked as well in the application using ordinary linkers.
The point here is to try to hide some dependencies to the end-user.

opti-mix replied:
> I know that on your part is harder to integrate this mechanism because you have a different build system.

It is much harder on our side using Buck and FB internal build configurations. Even without that, I'm still concerned about playing such games with CMake. 

> The point here is to try to hide some dependencies to the end-user.

I'm still not sure why you need so much CMake magic to achieve this "hiding" objective. If you have a pre-defined set of object files you want to always put into the archive for your backend, why not initialize/set the CLI flag with some default paths to your backend-specific/target-specific object files when you initialize your backend, e.g. in the constructor. This way the user does not need to know about this and it is hidden, but does not require any CMake tricks. More over, other backends derived from the LLVMBackend can customize this behavior by providing their own lists of object files, etc. It seems more modular to me. WDYT?

mciprian13 replied:
@opti-mix I cannot just set the CLI flag to some hard-coded path because I cannot guarantee the file will be at some given location. The CMake magic I do here is 99% the same as what we do with the libjit when serializing the bytecode as arrays and embedding them in the Glow binary. I don't see why we cannot do the same with this also.

opti-mix replied:
@mciprian13 

> I cannot just set the CLI flag to some hard-coded path because I cannot guarantee the file will be at some given location. 

Just checking: Would it work if you'd use a relative path (from e.g. some environment variable) instead of an absolute path and then resolve it at runtime? Something like "$LIBJIT_ROOT/objs/myobject1.o". You could then set this environment variable from shell or you could use a command-line flag to set this root directory.

> The CMake magic I do here is 99% the same as what we do with the libjit when serializing the bytecode as arrays and embedding them in the Glow binary. I don't see why we cannot do the same with this also.

I'm not a big fan of this part either ;-) And implementing this libjit serialization into bytecode and then packaging as array was quite a pain to implement using buck :-D

Another question is: IIRC, libjit is compiled by each LLVMBackend-derived backend separately (e.g. by CPUBackend). It is not compiled as part LLVMIRCodeGen. And you PR touches BundleSaver, which is part of the LLVMIRCodeGen. But what if multiple backends are using LLVMIRCodeGen? They should not all use the same BundleSaver with a specific version of the embedded object file. They probably want to use their own versions/sets of object files, which could be target/backend specific.

mciprian13 replied:
Well if this mechanism is not something you want I'll have to close this PR and merge it internally on our side. I would have preferred for you to except it in one form or another cuz otherwise it will become a pain to maintain it on our side. We need this mechanism ASAP in this exact form.

Just so you know what we do at NXP is not use Glow for ourselves but optimize it and package for others. So adding extra complexity to the usability of the Glow AOT tools (by relying on environment variables, hardcoded paths, etc) is not something we want. With this object archiving mechanism we just want to hide from the user the process of archiving additional objects (compiled with other compilers than clang) to the final object while being totally hidden from the user.

Can we add this mechanism only on the community part of Glow without adding it in your internal build system (if not of interest)? Maybe I can put it under some compile switches so that is build and ran only on github but not internally on your side? This way others might use this feature as well. WDYT?

opti-mix replied:
@mciprian13 

> Can we add this mechanism only on the community part of Glow without adding it in your internal build system (if not of interest)? Maybe I can put it under some compile switches so that is build and ran only on github but not internally on your side? This way others might use this feature as well. WDYT?

Yes, I guess this could be an option. 

But even in this case, I'd still highly recommend not changing the LLVMIRCodeGen (i.e. storing the object files under this directory) due to the reasons I described above. Try to move those object files into backends, so that different backends can bundle their own object files. After all, we do a similar trick for libjit, which lives inside backends.

opti-mix commented on lib/LLVMIRCodeGen/BundleSaver.cpp:
```diff
@@ -450,6 +457,66 @@ void BundleSaver::emitSymbolTable() {
                            irgen_->getBundleName() + "SymbolTable");
 }
 
+void BundleSaver::addBundleObjects(
```
The name `addBundleObjects` does not quite reflect what you do here, IMHO. You create an archive/static library. Maybe the name should be "createBundleArchive" or something like that?

mciprian13 replied:
Renamed it to `createBundleArchive` as you suggested.

opti-mix commented on lib/LLVMIRCodeGen/libobj/Readme.md:
```diff
@@ -0,0 +1,19 @@
+# Archiving additional objects within the bundle object
+
+All the object files (.o) stored in this directory will be automatically
```
Could you also provide the motivation for this feature and why it is useful?

mciprian13 replied:
I added some a description. I paste it here:

This is useful for the LLVM based backends in the following scenario:
- A LLVM backend generates a bundle with external calls to functions which are not part
  of LIBJIT. For example this is the case for some kernel libraries which are pre-compiled
  with other compilers than Clang (the one used by default by Glow) for better performance.
- This means that the bundle generated by Glow is not self contained and requires the
  respective library/objects to be provided at link time.
- For a better user experience we can choose not to delegate to the user the responsability
  of linking the additional libraries/objects during link-time but rather make Glow capable
  to archive the additional required objects files to the final bundle such that the bundle
  is self contained and the dependency is hidden from the user. During LLVMIRGen we can specify
  to add an extra object file to the bundle by using the function:
  ```
  void LLVMIRGen::addBundleObject(llvm::StringRef objectName);
  ```
  The above function is provided with the object name which is assumed to be present during Glow
  build time in this directory.


More specifically, for our internal use case:
- We currently have some ARM libraries integrated in Glow as source code in LIBJIT which is pre-compiled with Clang to LLVM-IR using the default LIBJIT flow.
- We noted that we get a better performance if we use other compiler (e.g. ARM GCC).
- What we want to do is to pre-compile the library using other compiler to an object file and provide it as dependency to the bundle at link-time (the bundle this time will only call external functions which are not part of LIBJIT but part of the external object file)
- We want this process to be hidden from the user because:
  - We do not want to have an additional step in which we instruct the user to provide the extra object file
  - We have project examples/documentation we do not want to update
  - We don't want the break the current way of using Glow just because the bundle has some external dependencies which Glow could provide

opti-mix reviewed (commented):
@mciprian13 Thanks for your PR! See my comments below. I'm a bit skeptical about the build-system based approach you've taken.

mciprian13 commented:
@opti-mix Any other opinions on this PR? I'm open in making this more flexible if need be but more or less this is the functionality I need. Actually I would be glad if we could review and freeze this PR even if we don't land it (guess it will take some time to write the CMake correspondent in your buck build system) so that I can merge it locally on my side to use it. Thanks!

mciprian13 commented:
@opti-mix I refactored this feature in the following way:
- As suggested, I moved the ownership of the object from BundleSaver to Backends themselves (similar to libjit) with the difference that I added a function called `getObjectRegister()` in the base class `Backend`. I think that serializing might be useful not just for the LLVM based backends but for all backends in general where it might have other purposes than store object code.
- I removed the folder `libobj` from LLVMIRCodeGen and added a folder `libobj_cpu` in the CPU backend. Currently this feature is enabled only in the CPU backend but other backends can enable this simply with a couple of lines of code.
- I had to modify slightly the constructor of LLVMIRGen such that the backend provides not only the libjit code but also the object register.
As you mentioned, we could simply disable this feature on your side by simply deactivating the lines of code from the end of 'CPUBackend.cpp'. Can you have another look and tell me what you think? Thanks!

opti-mix commented on lib/Backends/CPU/CPUBackend.cpp:
```diff
@@ -147,3 +147,14 @@ bool CPUBackend::canDoIndexTypeDemotion(
   return fromTy == ElemKind::Int64ITy && toTy == ElemKind::Int32ITy;
 }
+
+/// \returns an array of raw objects which are statically allocated and
```
This is an override and probably just copies the comment from a parent method? If so, let's not repeat it here.

mciprian13 replied:
Removed the comment. Done.

opti-mix commented on lib/Backends/CPU/CPUBackend.h:
```diff
@@ -67,6 +67,14 @@ class CPUBackend : public LLVMBackend {
   canDoIndexTypeDemotion(ElemKind fromTy, ElemKind toTy,
                          PrecisionConfiguration &precConfig) const override;
+
+  /// \returns an array of raw objects which are statically allocated and
```
This is an override and probably just copies the comment from a parent method? If so, let's not repeat it here.

mciprian13 replied:
Removed the comment. Done.

opti-mix commented on lib/Backends/CPU/CPULLVMIRGen.h:
```diff
@@ -29,7 +29,8 @@ class CPULLVMIRGen : public LLVMIRGen {
   /// Ctor.
   explicit CPULLVMIRGen(const IRFunction *M, AllocationsInfo &allocationsInfo,
-                        std::string mainEntryName, llvm::StringRef libjitBC);
+                        std::string mainEntryName, llvm::StringRef libjitBC,
```
Maybe it is better to define another additional constructor which takes an object registry and leave the existing one as is to avoid breaking any existing code? Existing constructor is then equivalent to calling this one with an empty object registry, right?

mciprian13 replied:
Restored the old constructor for CPULLVMIRGen. Now we have both constructors.
Yes, existing constructor will leave the object registry empty.
Done.

opti-mix commented on lib/Backends/CPU/CMakeLists.txt:
```diff
@@ -79,3 +80,22 @@ target_link_libraries(CPUBackend
 set(linked_backends ${linked_backends} CPUBackend PARENT_SCOPE)
+
+# List the object files from "libobj_cpu" directory.
+FILE(GLOB objects "${CMAKE_CURRENT_LIST_DIR}/libobj_cpu/*.o")
```
Could you clarify where this directory `libobj_cpu` with object files lives? In the source tree? If so, does it mean that all the *.o files are under source control? It would be a bit unusual. Is it a good idea to have those object files under source control inside Glow? Maybe they could live somewhere outside and some command-line parameter for CMake would provide a path to them, so that they can be found here and added to the bundle?

mciprian13 replied:
The folder `libobj_cpu` currently lives in the source tree along with `libjit_cpu`.
Currently we have no objects in that folder (only a `Readme.txt`) so we can say this code serves as an example of how backends can serialize and bundle their own objects. But backends (either public or private) can choose a different folder where to store their objects, for example they can store them externally and add a CMake option to provide at Glow build-time the external folder which contains the objects. Each backend can specialize/modify this as they please.

opti-mix reviewed (commented):
@mciprian13 This is looking much better! Thanks for your re-factoring of the PR. Please see my comments below.

mciprian13 commented on include/glow/Backend/Backend.h:
```diff
@@ -276,6 +276,16 @@ class Backend : public Named {
+  /// other compilers than clang/LLVM. The raw buffers are encoded as type
+  /// MemoryBufferRef which stores for each buffer a name, a raw pointer
+  /// and a size (in bytes).
+  virtual llvm::ArrayRef<llvm::MemoryBufferRef> getObjectRegister() const {
```
Done.

mciprian13 commented on include/glow/LLVMIRCodeGen/BundleSaver.h:
```diff
@@ -61,6 +61,14 @@ class BundleSaver {
+  /// \p bundleObjectRegister.
+  virtual void createBundleArchive(
+      llvm::StringRef bundlePath,
+      llvm::ArrayRef<llvm::MemoryBufferRef> bundleObjectRegister,
```
Done.

mciprian13 commented on include/glow/LLVMIRCodeGen/LLVMIRGen.h:
```diff
@@ -321,6 +330,17 @@ class LLVMIRGen {
+  /// \param objectRegister array with additional objects for code generation.
+  explicit LLVMIRGen(const IRFunction *M, AllocationsInfo &allocationsInfo,
+                     std::string mainEntryName, llvm::StringRef libjitBC,
+                     llvm::ArrayRef<llvm::MemoryBufferRef> objectRegister);
```
Done. Renamed everywhere "register" to "registry".

mciprian13 commented on include/glow/LLVMIRCodeGen/LLVMIRGen.h:
```diff
@@ -450,6 +470,15 @@ class LLVMIRGen {
   /// following the standard content produced by the bundle saver.
   virtual std::string getBundleHeaderExtra() const;
+  /// \returns the object register for this code generator instance.
+  llvm::ArrayRef<llvm::MemoryBufferRef> getObjectRegister() const;
```
Made them virtual. Done.

opti-mix commented:
@mciprian13 Thanks for addressing the review comments. Let's see if this PR builds on our internal CI.

mciprian13 commented:
@opti-mix Thanks for reviewing it. Most likely it will not build because I need to add a macro switch at the end of `CPUBackend.cpp` something along the lines of:
```
#if !FACEBOOK_INTERNAL
#include "cpuObjectRegistry.h"
llvm::ArrayRef<llvm::MemoryBufferRef> CPUBackend::getObjectRegistry() const {
  return cpuObjectRegistry;
}
#endif
```

opti-mix commented on lib/Backends/CPU/CPUBackend.cpp:
```diff
@@ -147,3 +147,8 @@ bool CPUBackend::canDoIndexTypeDemotion(
   return fromTy == ElemKind::Int64ITy && toTy == ElemKind::Int32ITy;
 }
+
+#include "cpuObjectRegistry.h"
```
OK. This include (and probably the function below it) is an issue for our build system that is not CMake based. Can you try to hide this behind some kind of conditional compilation? This feature should be enabled with a special CMake -D... parameter or something like that. By default, this include should not happen and no information from this include should be used.

mciprian13 replied:
@opti-mix Can we use the approach described above by wrapping under:
```
#if !FACEBOOK_INTERNAL
...
#endif
```
Isn't FACEBOOK_INTERNAL macro enabled by default in your builds?

opti-mix replied:
Yes, FACEBOOK_INTERNAL should work as well I guess.

opti-mix commented on lib/Backends/CPU/CPUBackend.cpp:
```diff
@@ -147,3 +147,10 @@ bool CPUBackend::canDoIndexTypeDemotion(
+
+#if !FACEBOOK_INTERNAL
+#include "cpuObjectRegistry.h"
+llvm::ArrayRef<llvm::MemoryBufferRef> CPUBackend::getObjectRegistry() const {
```
I think this function should always have an implementation, even under FB_INTERNAL as the class defines this method as existing. But the body of this method should return an empty arrayref and not use cpuObjectRegistry.

Or alternatively, instead of getting cpuObjectRegistry from the include file above, you can define it as a static empty array and return it here.

mciprian13 replied:
You're right. What about now?

opti-mix commented:
@mciprian13 The latest version looks good on our internal CI. There is one minor issue preventing it from landing, which is unrelated to your changes. Once it is fixed, I'll merge this PR. Thanks again for this feature contribution!
